SCRUM-242 feat: implement layering and naming

gdaegeun539 opened the pull request:
### Please check if the PR fulfills these requirements
- [x] The commit message follows our guidelines
<!-- - [ ] Tests for the changes have been added (for bug fixes / features) ~~  -->
<!-- - [ ] Docs have been added / updated (for bug fixes / features) -->


## What kind of change does this PR introduce?
- Add feature


# What is the current behavior?
>  You can also link to an open issue here

SCRUM-242


# What is the new behavior (if this is a feature change)?
- 데이터 레이어에 대한 패키지/파일 틀과 네이밍을 추가합니다.
  - 인증/로그인/로그아웃 기능에 대한 파일을 이용합니다.
  - 세부적인 클래스 요소(파라미터/프로퍼티/메소드)는 실제 기능을 만들 때 변경해도 될 것이라 생각합니다.
- 하단에 첨부된 다이어그램을 생각하며 제작했습니다.


## Does this PR introduce a breaking change? (What changes might users need to make in their application due to this PR?)
- 아래와 같이 패키지 분류와 네이밍을 지정했는데, 이견 있으실 경우 리뷰 부탁드립니다.
  - data layer의 파일들은 `core.data.repository`, `core.data.service`로 패키지를 지정했습니다.
  - 인터페이스/구현체는 `XXX`/`XXXImpl`로 명명했습니다.
  - (자체)서버 API 통신 인터페이스/클래스는 `XXXApiService`, SDK 사용 클래스는 `XXXSdkSerivce`로 명명했습니다.
- 최대한 pokedex와 nia를 참고해 안드로이드(+멀티모듈 확장)에 맞게 명명 및 배치하려 노력했습니다.


## ScreenShots
<p>
  <img src="https://github.com/user-attachments/assets/8be2fcdb-080f-4c6e-824b-54f9178a3a73" alt="로그인만 생각한 레이어 틀" width="600" />
</p>


## Other information:
- hilt 설정 및 레이어 간 di는 실제 구현 시 진행하겠습니다.
- retrofit 설정은 서버간 api 통신을 진행할 때 구현하겠습니다.

gdaegeun539 committed:
feat: add login viewmodel frame

gdaegeun539 committed:
feat: add oauth provider enum

gdaegeun539 committed:
feat: add auth repo, api/sdk service frame

hyunjung-choi commented on app/src/main/java/com/lyrics/feelin/core/data/repository/AuthRepository.kt:
### Repository 인터페이스/구현체 분리 제거
구글 권장 방식에서는 Repository를 인터페이스로 분리하지 않고 클래스 하나로 사용합니다.

**수정 전:**
- `AuthRepository` (interface)
- `AuthRepositoryImpl` (implementation)

**수정 후:**
- `AuthRepository` (class)

지금 구현하신 방식은 클린 아키텍처에 좀 더 가깝다고 볼 수 있습니다 !

gdaegeun539 replied:
아 그러네요 앱 아키텍처에서는 바로 구현체를 써도 상관이 없네요.

hyunjung-choi commented on app/src/main/java/com/lyrics/feelin/core/data/service/AuthApiService.kt:
### 패키지명 변경: service → datasource

`Service`는 주로 Retrofit API 인터페이스를 지칭합니다.

```
data/
  └── datasource/
      ├── remote/
      │   ├── AuthRemoteDataSource.kt    // Retrofit 래핑
      │   └── AuthApiService.kt          // Retrofit interface
      └── sdk/
          ├── KakaoAuthDataSource.kt     // Kakao SDK 래핑
          └── GoogleAuthDataSource.kt    // Google SDK 래핑
```

가능하면 `local` `remote` `sdk` 이렇게 패키징해서 데이터 출처도 명확하게 하는게 좋을 것 같습니다 !

gdaegeun539 replied:
`XXXService`가 retrofit 인터페이스임을 명확히 알지 못하고 정했던 네이밍이라, 이대로 진행하는게 더 맞을 듯 해요.

해당 내용을 알지 못하고 Claude 4 Sonnet이랑 논의했을 때 내용은 아래와 같아요. (사용하지 않을 겁니다.)

제안하신 네이밍 컨벤션 xxxSdkService와 xxxApiService는 매우 훌륭한 아이디어입니다! 이에 대해 자세히 분석해드리겠습니다.
🎯 제안하신 네이밍 컨벤션 분석
✅ 장점들
명확한 구분: SDK vs API가 명확히 구분됨
일관성: 일관된 네이밍 패턴으로 예측 가능
확장성: 새로운 SDK나 API 추가 시 쉽게 적용
가독성: 코드 읽기 시 역할이 즉시 이해됨

hyunjung-choi commented on app/src/main/java/com/lyrics/feelin/core/data/repository/AuthRepository.kt:
```diff
@@ -0,0 +1,9 @@
+import com.lyrics.feelin.core.domain.model.OAuthProvider
+
+interface AuthRepository {
+    fun login(provider: OAuthProvider)
```
추후 `suspend fun`으로 비동기 처리가 구현되는거죠 . . ?

gdaegeun539 replied:
아 맞아요. 단순히 스펙만을 정의한 거라 비동기 처리까지는 생각하지 않았습니다.

hyunjung-choi commented on app/src/main/java/com/lyrics/feelin/core/data/repository/AuthRepositoryImpl.kt:
```diff
@@ -0,0 +1,27 @@
+
+class AuthRepositoryImpl(
+    @Suppress("UnusedPrivateMember") // TODO(@이대근): 구현 중 제거할 것. 2025.09.29.
+    private val kakaoOAuthSdkService: KakaoOAuthSdkService,
```
### DataSource 레이어 추가

Repository가 SDK를 직접 호출하지 말고, DataSource를 통해 접근하도록 해서
비즈니스 로직 조합만 담당하게 하는 것이 어떨까요 !?
(Service를 DataSource로 바꾸는 과정에서 하면 좋을 듯 합니다 !)

gdaegeun539 replied:
위 패키지 구조대로 SdkService를 AuthDataSource로 바꾸는 과정에 자연스럽게 진행하면 좋을 것 같아요. 의견 감사합니다!
(그리고 TODO 주석은 `@Supress` 어노테이션만 언급한 것이긴 한데, 조금 모호하게 적었네요.)

gdaegeun539 replied:
이렇게 하게 되면 도메인 레이어가 불필요할 때 까지는 한 레포지토리 레이어 내에서 약간의 도메인 로직을 가지게 되네요.
어쨌든 구글 앱 아키텍처가 불필요한 도메인 보일러플레이팅을 하지 않기 위한 것이니 이대로 진행해볼게요.

hyunjung-choi reviewed (approved):
대근님 너무 너무 고생하셨습니다 ! 🥲

제 리뷰가 정답은 아니라고 생각합니다 !
처음부터 완벽한 아키텍처를 설계하는 것 보다 **필요에 따라 점진적으로 개선하는 것**이 맞다고 생각하기 때문에 참고만 해주시고,
실제로 코드를 작성하시면서 "이 부분이 불편하네" 싶을 때 개선을 고려하는 것도 괜찮을 것 같습니다.

리뷰하면서 참고한 자료 공유드립니다.
- [Data Layer 가이드](https://developer.android.com/topic/architecture/data-layer)
- [Repository 코드랩](https://developer.android.com/codelabs/basic-android-kotlin-compose-add-repository)

gdaegeun539 commented:
> 제 리뷰가 정답은 아니라고 생각합니다 ! 처음부터 완벽한 아키텍처를 설계하는 것 보다 **필요에 따라 점진적으로 개선하는 것**이 맞다고 생각하기 때문에 참고만 해주시고, 실제로 코드를 작성하시면서 "이 부분이 불편하네" 싶을 때 개선을 고려하는 것도 괜찮을 것 같습니다.

굉장히 중요한 댓글을 써 주셨어요. 부딪히는게 무서워서 설계만 하고 있으면 발전하지 않다 보니 진행하면서 개선점을 찾는게 굉장히 중요한데 이 부분을 간과한 것 같아요.
리뷰 진행해주셔서 감사합니다!

gdaegeun539 commented:
```markdown
data/
  └── datasource/
      ├── remote/
      │   ├── AuthRemoteDataSource.kt    // Retrofit 래핑
      │   └── AuthApiService.kt          // Retrofit interface
      └── sdk/
          ├── KakaoAuthDataSource.kt     // Kakao SDK 래핑
          └── GoogleAuthDataSource.kt    // Google SDK 래핑
```
![구글 앱 아키텍처에 맞게 바꾼 로그인만 생각한 레이어 틀](https://github.com/user-attachments/assets/9867f7af-a851-4115-adac-9787d455097e)

데이터 레이어만 위와 같게 변경한 후 바로 머지하겠습니다.

(수정 2025.10.02 17:06)
스크린샷에 구글 sdk 래퍼 클래스명이 변경되지 않았는데, `GoogleAuthDataSource`로 폴더 구조의 이름과 동일하게 진행하겠습니다.

gdaegeun539 committed:
feat: modify data layer frame

- change service naming to datasource
- add remoteDataSource to wrap retrofit service
- remove repository interface to fit the Google app architecture
